packages/blocks/src/types.ts

```diff
@@ -0,0 +1,1504 @@
+	 *
+	 * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#script}
+	 */
+	script?: string;
+
+	/**
+	 * Block type editor style definition.
+	 * It will only be enqueued in the context of the editor.
+	 *
+	 * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#style}
+	 */
+	style?: string;
+
+	/**
+	 * Block styles.
+	 *
+	 * @see {@link https://wordpress.org/gutenberg/handbook/extensibility/extending-blocks/#block-style-variations}
+	 */
+	styles?: BlockStyle[];
+
+	/**
+	 * Optional block extended support features.
+	 *
+	 * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/ Block Supports} on developer.wordpress.org
+	 */
+	supports?: BlockSupports;
+
+	/**
+	 * The gettext text domain of the plugin/block.
+	 *
+	 * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#text-domain}
+	 */
+	textdomain?: string;
+
+	/**
+	 * This is the display title for your block, which can be translated
+	 * with our translation functions.
+	 */
+	title: string;
+
+	/**
+	 * Block transformations.
+	 */
+	transforms?:
+		| {
+				/**
+				 * Transforms from another block type to this block type.
+				 */
+				from?: Transform< Attributes >[];
+				/**
+				 * Transforms from this block type to another block type.
+				 */
+				to?: Transform[];
+				/**
+				 * TODO Undocumented
+				 *
+				 * Used in [factory.js](./api/factory.js)
+				 */
+				supportedMobileTransforms?: string[];
+		  }
+		| undefined;
+
+	/**
+	 * Array of the names of context values to inherit from an ancestor
+	 * provider.
+	 *
+	 * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#context}
+	 */
+	usesContext?: string[];
+
+	/**
+	 * The current version number of the block, such as 1.0 or 1.0.3.
+	 *
+	 * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#version}
+	 */
+	version?: string;
+
+	/**
+	 * Sets attributes on the topmost parent element of the current block.
+	 */
+	getEditWrapperProps?(
+		attrs: Attributes
+	): Record< string, string | number | boolean >;
+
+	/**
+	 * TODO Undocumented, but used in core.
+	 *
+	 * @see `@wordpress/block-library/src/paragraph`
+	 */
+	merge?(
+		attributes: Attributes,
+		attributesToMerge: Attributes
+	): Partial< Attributes >;
+
+	/**
+	 * TODO Undocumented
+	 *
+	 * @internal
+	 */
+	__experimentalLabel(
+		attributes: BlockAttributes,
+		settings: { context: 'accessibility' | 'visual' } // TODO find all context types
+	): string;
+}
+
+/**
+ * The raw structure of a block includes its attributes, inner
+ * blocks, and inner HTML. It is important to distinguish inner blocks from
+ * the HTML content of the block as only the latter is relevant for block
+ * validation and edit operations.
+ *
+ * @public
+ */
+export type BlockRaw = {
```

Comment:
this is what we have come to call the "block node," or in a different description, the result of the "stage-1 parse of a block"

Reply:
If you could adjust the names to fit I would appreciate it. I was doing my best to marry the DefinitelyTyped definitions with what was in the JSDocs, but I'm still getting familiar with everything. That or as a new developer the lifecycle of a block is still a bit of a mystery. I understand it better by looking thought this package, but is there a document that explains the concepts and terms of the block lifecycle?

Reply:
unfortunately I don't think this is in the code. I started some work #15674 but that's out of date now, and we have held informal conversations around these terms.

the existing references are in this file and we can either choose to keep `RawBlock` (`BlockRaw` is a bit awkward linguistically) or adopt `BlockNode`, which is how I believe we want to refer to this going forward. This term is used in the `children` file.

I realize we talk about `blockNode` in the RichText/DOM stuff, but these are also two different kinds of "node" in play. not sure the best resolution; could be to leave the name clash; could be to pick one and change the other.

Reply:
I'll take a look at that PR, even if outdated, I'm sure has helpful information.

Reply:
This is related to https://github.com/WordPress/gutenberg/pull/48604#discussion_r1142826547. I prefer the name `BlockNode` over `BlockRaw` or `RawBlock` for this stage in the block lifecycle.